py/redrock/utils.py

```diff
@@ -216,22 +216,51 @@ def distribute_work(nproc, ids, weights=None, capacities=None):
     return dist
 
 
-def transmission_Lyman(zObj,lObs):
+def transmission_Lyman(zObj,lObs, use_gpu=False):
     """Calculate the transmitted flux fraction from the Lyman series
     This returns the transmitted flux fraction:
     1 -> everything is transmitted (medium is transparent)
     0 -> nothing is transmitted (medium is opaque)
+
+    This method will handle 3 options:
+    1 -> For GPU mode, zObj is an array of all z for a template and the return
+    value will be a cupy array (nz x nlambda)
+    2-> In CPU mode, it can auto-detect if zObj is a numpy array and if so,
+    again, all z will be processed as a vector and the return value
+    will be a numpy array (nz x nlambda)
+    3-> For legacy, it is still supported to pass zObj as a float and
+    in this case, the return value will be a 1-d numpy array (nlambda).
```

Comment:
Do the unit tests cover these 3 cases, including that the return type is as expected?  cupy vs. numpy array propagation is still a bit of a mystery to me (e.g. does `T = np.ones_like(lRF)` return a cupy array if `lRF` is a cupy array?) so it could be helpful to explicitly confirm that it does what we think and protect against future updates that might accidentally trigger a GPU -> CPU array transition.

Reply:
Yes, the unit tests cover all 3 cases verifying that the results match (or are `allclose`) for all three.  I don't explicitly check the output return type but I can add that easily.  In the method, I explicitly cast `zObj` (and thus `lRF`) as numpy or cupy array depending on the `use_gpu` flag.

And yes, `T = np.ones_like(lRF)` will return a cupy array if `lRF` is a cupy array and a numpy array if it is a numpy array.

Reply:
Minor type surprise: `transmission_Lyman(zObs, lObs, use_gpu=True)` returns a cupy array if `zObs` is an array (expected), but returns a numpy array if `zObs` is a scalar.  I was expecting cupy vs. numpy array to depend only upon `use_gpu`, not upon the dimensionality of `zObs`.  I'm not sure it matters, but if it is a simple fix to ensure that use_gpu=True always returns cupy arrays, that seems more consistent.

Is this supposed to work?
```
transmission_Lyman(cupy.asarray(zObj), lObs)
```
Passing in zObj as a cupy array without explicitly setting use_gpu=True raises an exeception
```
File /global/common/software/desi/users/sjbailey/redrock/py/redrock/utils.py:262, in transmission_Lyman(zObj, lObs, use_gpu)
    260         return None
    261     lObs = tile(lObs, (zObj.size, 1))
--> 262     lRF = lObs/(1.+asarray(zObj)[:,None])
    263 T = np.ones_like(lRF)
    264 for l in list(Lyman_series.keys()):

File cupy/_core/core.pyx:1381, in cupy._core.core.ndarray.__array__()

TypeError: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.
```

This might not matter, but in the spirit of making it as transparent as possible whether we are passing around numpy vs. cupy arrays, let's fix this if it is trivial to do so (and if it isn't easy, it's ok to wait and see if this becomes a real problem or not).